### ISSUE TITLE
Unable to execute method="SR1" 

### ISSUE THREAD

**Issue by Haimonti** [jlmelville/mize#2]
I have been trying to execute the “SR1” algorithm, as described in Nocedal and Wright, and came upon the implementation in the “mize” package in R.

I tried to use version 0.2.4 of mize, but without any luck. Please refer to the error below.

res.loss1<-mize(c(1,1,1),list(fn=loss.fn, gr=grad.fn, hs=hess.fn), method="SR1" )
Error in match.arg(tolower(method), c("sd", "newton", "phess", "cg", "bfgs",  : 
 'arg' should be one of “sd”, “newton”, “phess”, “cg”, “bfgs”, “l-bfgs”, “nag”, “momentum”, “dbd”

In the above, I was simply testing a squared loss, with its corresponding gradient and hessian on a small, binary dataset. 

Some more commands from the R console:

library("mize")
% Read in the data and get relevant features
> A<-read.csv('/Users/haimontidutta/Research/DDM/DistSVM_QuasiNewton/Code/RCode/iris_new.csv',header=TRUE,sep=",")
% I use only two features and the class for testing
> subA=as.matrix(A[,1:2])
> cl=A[,5]
> dat=cbind(subA,cl)
> dat=as.data.frame(dat)
> names(dat)<-c("Sepal_Length","Sepal_Width","Class")

%% Definitions of loss functions, gradients and hessians -- I use squared loss
sqloss_fn<- function(wt,d)
{
sum((wt[1] + wt[2] * d[,1] + wt[3] * d[,2] - d[,3])^2)
}

sqloss_gr<-function(wt,d)
{
 numRows<-nrow(d)
  pred<-d[,3] - (wt[1] + wt[2] * d[,1] + wt[3] * d[,2])
  tmp1<-cbind(1,d[,1],d[,2])
   tmp2<- (- 2 * 1/(numRows) * t( t(tmp1) * pred))
   colSums(tmp2)
} 

sqloss_hs<-function(d)
{
subA<-d[,1:2]
2 * t(subA) %*% (subA)
}

loss.fn<-function(par)
{
d<-dat
wt<-par
sqloss_fn(wt,d)
}

grad.fn<-function(par)
{
d<-dat
wt<-par
sqloss_gr(wt,d)
}

hess.fn<-function(par)
{
d<-dat
sqloss_hs(d)
}

%% The execution with method="BFGS" produces the following output
> res.loss<-mize(c(1,1,1),list(fn=loss.fn,gr=grad.fn, hs=hess.fn), method="BFGS")
> res.loss
$f
[1] 6.156157

$nf
[1] 30

$ng
[1] 30

$par
[1]  0.6017066  0.4855947 -0.8915795

$iter
[1] 5

$terminate
$terminate$what
[1] "step_tol"

$terminate$val
[1] 0

%% When I try to execute with SR1 it produces this error indicated above
res.loss1<-mize(c(1,1,1),list(fn=loss.fn, gr=grad.fn, hs=hess.fn), method="SR1" )
Error in match.arg(tolower(method), c("sd", "newton", "phess", "cg", "bfgs",  : 
  'arg' should be one of “sd”, “newton”, “phess”, “cg”, “bfgs”, “l-bfgs”, “nag”, “momentum”, “dbd”

Also, sessionInfo() is as follows:

> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.15.7 (unknown)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mize_0.2.4

loaded via a namespace (and not attached):
 [1] Matrix_1.2-12    magrittr_2.0.1   R6_2.5.0         generics_0.1.0   tools_3.3.1      whisker_0.3-2    base64enc_0.1-3  Rcpp_1.0.6      
 [9] reticulate_1.18  keras_2.3.0.0    tensorflow_2.2.0 grid_3.3.1       zeallot_0.0.6    jsonlite_1.5     tfruns_1.1       lattice_0.20-35 

I was wondering why I cannot execute method="SR1".  

I really appreciate the help with this.

**Comment by jlmelville** [comment]
@Haimonti I don't know what's going on. The error message quotes code that isn't in version 0.2.4 of mize (I downloaded the source package from CRAN to double check).

By replacing the data you are loading with the standard R `iris` dataset, I am able to run the following code on Windows and Linux versions of R, so this *should* be reproducible by copy and pasting into your R environment:

``` r
install.packages("mize")
#> Installing package into 'E:/dev/R/win-library/4.0'
#> (as 'lib' is unspecified)
#> package 'mize' successfully unpacked and MD5 sums checked
#> 
#> The downloaded binary packages are in
#>  C:\Users\jlmel\AppData\Local\Temp\RtmpcX0bAW\downloaded_packages
library("mize")
packageVersion("mize")
#> [1] '0.2.4'

A <- iris
subA <- as.matrix(A[, 1:2])
cl <- A[, 5]
dat <- cbind(subA, cl)
dat <- as.data.frame(dat)
names(dat) <- c("Sepal_Length", "Sepal_Width", "Class")

sqloss_fn <- function(wt, d)
{
    sum((wt[1] + wt[2] * d[, 1] + wt[3] * d[, 2] - d[, 3]) ^ 2)
}

sqloss_gr <- function(wt, d)
{
    numRows <- nrow(d)
    pred <- d[, 3] - (wt[1] + wt[2] * d[, 1] + wt[3] * d[, 2])
    tmp1 <- cbind(1, d[, 1], d[, 2])
    tmp2 <- (-2 * 1 / (numRows) * t(t(tmp1) * pred))
    colSums(tmp2)
}

sqloss_hs <- function(d)
{
    subA <- d[, 1:2]
    2 * t(subA) %*% (subA)
}

loss.fn <- function(par)
{
    d <- dat
    wt <- par
    sqloss_fn(wt, d)
}

grad.fn <- function(par)
{
    d <- dat
    wt <- par
    sqloss_gr(wt, d)
}

hess.fn <- function(par)
{
    d <- dat
    sqloss_hs(d)
}


res.loss.BFGS <-
    mize(c(1, 1, 1), list(fn = loss.fn, gr = grad.fn, hs = hess.fn), method =
             "BFGS")
res.loss.BFGS
#> $f
#> [1] 135.0844
#> 
#> $nf
#> [1] 25
#> 
#> $ng
#> [1] 25
#> 
#> $par
#> [1]  0.82258548 -0.04569195  0.46621622
#> 
#> $iter
#> [1] 3
#> 
#> $terminate
#> $terminate$what
#> [1] "step_tol"
#> 
#> $terminate$val
#> [1] 6.938894e-17

res.loss.SR1 <-
    mize(c(1, 1, 1), list(fn = loss.fn, gr = grad.fn, hs = hess.fn), method =
             "SR1")
res.loss.SR1
#> $f
#> [1] 134.3143
#> 
#> $nf
#> [1] 26
#> 
#> $ng
#> [1] 26
#> 
#> $par
#> [1]  0.8224732 -0.0432106  0.4614990
#> 
#> $iter
#> [1] 2
#> 
#> $terminate
#> $terminate$what
#> [1] "step_tol"
#> 
#> $terminate$val
#> [1] 1.927964e-16
```

<sup>Created on 2021-03-11 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

If you can confirm that you get an error when running this code with SR1, then the only things I can think of are:

* Reinstalling mize (and make sure it reinstalls without errors, and that `packageVersion("mize") == "0.2.4"`).
* Upgrading your version of R to 4.x (that shouldn't matter, but the Windows and Linux versions of R I am using are both on 4.0.x).

**Comment by Haimonti** [comment]
@jlmelville It worked! An upgrade of R to 4.x seemed to do the trick. Not sure why, though. 

Thanks for all the help with debugging this. I will close the issue.
